google/TensorNetwork
# Algebraic operation add( + ), sub( - ), mul( * ), div( / ) for BaseNode class

Fix #292

**[BaseNode class]** - add / sub / mul / truediv NotImplemented function Added

     [Node class] - add / sub / mul / truediv function added
     [CopyNode class] - overload the BaseNode add / sub / mul / truediv as NotImplemented

**[basebackend class]** - add / sub / mul / div NotImplemented function added

     [numpy / tensorflow / pytorch] - add / sub / mul / div function added
     [shell] - add / sub / div NotImplemented function added

Testing files
**[network_components_free_test]**

     Exception - Tensorflow is not tested when the operand is scalar
        1. Check add / sub / mul / div with int / float / Node
        2. Check implicit conversion
        3. Check the Type Error when type is not int / float / Node
        4. Check is the operand backend same
        5. Check is BaseNode has attribute _tensor

**[backend_test - numpy / tensorflow / pytorch]**

        check add / sub / mul / divide work for int / float / Node

TensorFlow doesn’t accept scalar value because `scalar value dtype` will be `32 bits` and `matrix` will be `64 bits`. There is no way to change the tensor’s dtype.

**Here is my suggestions while I think about doing this.**

1. **Add dtype attribute to BaseNode to handle TensorFlow**. 
TensorFlow treats the array as 64 bit and scalar as 32 bit. So we need to modify the BaseNode to accept the datatype.

2. **BaseNode tensor type check** 
There is no type check at the tensor itself so below throws no error.

tensor assigned for `str` type

    node = tn.Node(tensor='str')

But didn’t put protection checking the tensor type at the operator part. Because this responsibility is at tensor itself in my opinion. It will be good to put the protection when we put a value at the tensor. 

    node.tensor = value
